rust/pattern/variable/builder/thing_variable_builder.rs

```diff
@@ -36,52 +37,84 @@ pub trait ThingVariableBuilder {
         self,
         type_name: impl Into<String>,
         value: T,
-    ) -> Result<Variable, ErrorMessage>
+    ) -> Result<ThingVariable, ErrorMessage>
     where
         ErrorMessage: From<<T as TryInto<Value>>::Error>;
 
-    fn iid<T: TryInto<IIDConstraint>>(self, iid: T) -> Result<Variable, ErrorMessage>
+    fn iid<T: TryInto<IIDConstraint>>(self, iid: T) -> Result<ThingVariable, ErrorMessage>
     where
         ErrorMessage: From<<T as TryInto<IIDConstraint>>::Error>;
 
-    fn isa(self, isa: impl Into<IsaConstraint>) -> Result<Variable, ErrorMessage>;
+    fn isa(self, isa: impl Into<IsaConstraint>) -> Result<ThingVariable, ErrorMessage>;
 
-    fn eq(self, value: impl Into<Value>) -> Result<Variable, ErrorMessage>;
+    fn eq(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
```

Comment:
Always return a concrete struct instance in the builder functions, so that we can leverage static analysis and compile-time type checking and avoid dynamic type checks (e.g. `Variable` checking if it is a `Thing` or `Unbound` on an `isa()` call.